File: tests/xcalablemp/global-view/loop/C/reduction_mult.c
#include <xmp.h>
#include <stdio.h>  
#include <stdlib.h> 
static const int N=1000;
#pragma xmp nodes p(*)
#pragma xmp template t(0:N-1)
#pragma xmp distribute t(block) onto p
int a[N],sa=1;
double b[N],sb=1.0;
float c[N],sc=1.0;
int i,result=0;
#pragma xmp align a[i] with t(i)
#pragma xmp align b[i] with t(i)
#pragma xmp align c[i] with t(i)

int main(void)
{
#pragma xmp loop (i) on t(i)
  for(i=0;i<N;i++){
    if((i+1)%100 == 0){
      a[i] = 2;
      b[i] = 2.0;
      c[i] = 4.0;
    }else if((i+1)%100 == 50){
      a[i] = 1.0;
      b[i] = 0.5;
      c[i] = 0.25;
    }else{
      a[i] = 1;
      b[i] = 1.0;
      c[i] = 1.0;  
    }
  }

#pragma xmp loop on t(i) reduction(*:sa)
  for(i=0;i<N;i++)
    sa = sa*a[i];
  
#pragma xmp loop on t(i) reduction(*:sb)
  for(i=0;i<N;i++)
    sb = sb*b[i];
  
#pragma xmp loop on t(i) reduction(*:sc)
  for(i=0;i<N;i++)
    sc = sc*c[i];
  
  if((sa != 1024)||abs(sb-(double)1.0) >= 0.0000001 ||abs(sc-(float)1.0) >= 0.0001){
    result = -1;
  }
  
#pragma xmp reduction(+:result)
#pragma xmp task on p(1)
   {
     if(result == 0){
       printf("PASS\n");
     }
     else{
       fprintf(stderr, "ERROR\n");
       exit(1);
     }
   }
   return 0;
}

File: tests/XMP/global-view/loop/F/reduction_mult.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N)
!$xmp distribute t(block) onto p
  integer a(N), sa, result
  real*4 b(N), sb
  real*8 c(N), sc
!$xmp align a(i) with t(i)
!$xmp align b(i) with t(i)
!$xmp align c(i) with t(i)

!$xmp loop (i) on t(i)
  do i=1, N
     if(mod(i,100).eq.0) then
        a(i) = 2
        b(i) = 2.0
        c(i) = 4.0
     else if(mod(i,100).eq.50) then
        a(i) = 1.0
        b(i) = 0.5
        c(i) = 0.25
     else
        a(i) = 1
        b(i) = 1.0
        c(i) = 1.0
     endif
  enddo

  sa = 1
  sb = 1.0
  sc = 1.0

!$xmp loop (i) on t(i) reduction(*: sa, sb, sc)
  do i=1, N
     sa = sa*a(i)
     sb = sb*b(i)
     sc = sc*c(i)
  enddo

  result = 0
  if( sa .ne. 1024 .or. abs(sb-dble(1.0)) .gt. 0.0000001 .or. abs(sb-real(1.0)) .gt. 0.0001 ) then
     result = -1
  endif
  
!$xmp reduction(+:result)
!$xmp task on p(1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task
  
end program main
